Improve the performance of CAGRA new vector addition with the default params

enp1s0 opened the pull request:
This PR updates the default chunk size of the CAGRA graph extension and also adds a knob to control the batch size of the CAGRA searches run inside for better throughput.

The default chunk size was set to 1 in the current implementation because there is a potential problem with low recall when the chunk size is large, because no edges are made within nodes in the same chunk. However, as I have investigated, the low recall problem rarely occurs with large chunk sizes.

# Search performance

The performance was measured after applying a bugfix https://github.com/rapidsai/cuvs/pull/565

## degree = 32


![extend-ir0 9-degree32](https://github.com/user-attachments/assets/a5bb2fb6-8c12-49ad-b96a-1b384d79a96b)


(I don't know the reason the performance is unstable in NYTimes.)

## degree = 64
![extend-ir0 9-degree64](https://github.com/user-attachments/assets/8e926e1c-d772-4682-9419-9cc027f09d3f)

So I increase the default chunk size to the size of the new dataset vectors for better throughput in this PR. I also make public a knob to control the search batch size in the `extend' function to control the balance between throughput and memory consumption.

cjnolet commented:
/ok to test

tfeher reviewed (commented):
Thanks @enp1s0 for the PR! It looks good, but we should discuss whether we need a new parameter, or whether we could utilize the `workspace` memory resource for controlling the batch size.

The raft resource handle has a `workspace` memory resource. This is set up a an [rmm::mr::limiting_resource_adaptor](https://github.com/rapidsai/rmm/blob/d750a6e8578a4e61159faae59baeb4470e2ed81f/include/rmm/mr/device/limiting_resource_adaptor.hpp#L35). The intended usage of this is to query it's size, and adjust our internal batching parameter accordingly. 

In other algorithms we limit a batch size by querying the what is the available workspace size using [get_workspace_free_bytes](https://github.com/rapidsai/cuvs/blob/bd603a97ea7c095dd109c802387b33dc1d591b54/cpp/src/neighbors/ivf_pq/ivf_pq_search.cuh#L598)

Afterwards, we use the workspace memory resource to allocate the temporary buffers:
https://github.com/rapidsai/cuvs/blob/bd603a97ea7c095dd109c802387b33dc1d591b54/cpp/src/neighbors/ivf_pq/ivf_pq_search.cuh#L678-L684

Users who need fine control on temporary allocations can set the workspace allocator explicitly:
https://github.com/rapidsai/cuvs/blob/bd603a97ea7c095dd109c802387b33dc1d591b54/examples/cpp/src/cagra_example.cu#L76

Note that by default the allocation limit is 1/4th of GPU Total memory, which is potentially much larger then the default for `max_working_device_memory_size_in_megabyte` introduced` in this PR.

enp1s0 commented:
Thank you @tfeher for the review. I changed the code to use `get_workspace_free_bytes` and removed `max_working_device_memory_size_in_megabyte`. The original code has already used `raft::resource::get_workspace_resource`, so I just changed the max batch size calculation.

tfeher reviewed (changes requested):
Thanks Hiroyuki for the update! I have one more question below.

enp1s0 commented on cpp/src/neighbors/detail/cagra/add_nodes.cuh:
```diff
@@ -68,7 +69,12 @@ void add_node_core(
 
-  const std::size_t max_chunk_size = 1024;
+  const std::size_t data_size_per_vector =
+    sizeof(IdxT) * base_degree + sizeof(DistanceT) * base_degree + sizeof(T) * dim;
```
Thank you @tfeher for the suggestion. Yes, the term for `batch_load_iterator` should be added. I fixed the code.

tfeher reviewed (approved):
Thanks @enp1s0 for the updates, the PR looks good to me!

cjnolet commented:
/merge
